コメント投稿機能(Comments.index)

tattu501 opened the pull request:
#何か
掲示板詳細ページにコメント投稿機能を追加
<img width="1280" alt="スクリーンショット 2020-11-03 18 33 43" src="https://user-images.githubusercontent.com/61368201/97969484-76b15380-1e03-11eb-8081-8ecbafcb8e89.png">

tattu501 committed:
commentsコントローラを作成

commentsコントローラの作成により、必要なファイルが作られた。

tattu501 committed:
commentモデルを生成

commentモデル生成に伴い、必要なファイルが作成された

tattu501 committed:
routes.rbにresources :commentsを追加

コメント投稿機能を作成するため

tattu501 committed:
comments.controller.rbにcreateアクションを追加

コメント投稿機能を作成するため、createアクションを追加
※メソッドの定義のみで、まだ中身はありません、

tattu501 committed:
コンフリクト解消のため、_header.html.erbを修正

masterとのコンフリクトを解消するため、内容をmasterに合わせる

tattu501 committed:
Merge branch 'posts.edit/delete' into comments.index

tattu501 committed:
schema.rbにcommentモデルを追加

rails db:migrateでcommentモデルを追加

tattu501 committed:
Merge branch 'header' into comments.index

tattu501 committed:
routes.rbにcommentsのルーティングを追加

resources :postsの中にresources :comments, only: [:create]を追加

tattu501 committed:
comments_controller.rbにcreateアクションを追加

コメント登録用にcreateアクションを追加

tattu501 committed:
posts_controller.rbのshowアクションを修正

掲示板詳細ページにコメント一覧、および投稿フォームを作成するため

tattu501 committed:
comments.rbにコメントのバリデーションを追加

項目：コメント（comment）に以下のバリデーションを追加した。
・必須入力であること
・140字以下であること

tattu501 committed:
post.rbにhas_many :commentsを追加

postsとcommentを紐付けるため、post.rbにhas_many :commentsを追加

tattu501 committed:
user.rbにhas_many :commentsを追加

usersとcommentを紐付けるため、user.rbにhas_many :commentsを追加

tattu501 committed:
posts/index.html.erbにコメント投稿数を追加

posts/index.html.erbにコメント投稿数を表示するため、<%= post.comments.count %>を追加

tattu501 committed:
posts/show.html.erbにコメント一覧と投稿フォームを追加

コメント一覧と投稿フォームを追加

tattu501 committed:
app/views/layouts/application.html.erb.tmpを削除

不要なファイル：app/views/layouts/application.html.erb.tmpを削除

tattu501 committed:
app/assets/javascripts/comments.coffeeを削除

不要なファイル：app/assets/javascripts/comments.coffeeを削除

msgyu commented on app/assets/stylesheets/comments.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the comments controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
使用していないので削除

tattu501 replied:
上記ファイルを削除しました。

msgyu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,16 @@
+class CommentsController < ApplicationController
+  def create
```
`before_action :authenticate_user!`の追加をしておきましょう

tattu501 replied:
before_action :authenticate_user!を追加しました。

msgyu commented on app/views/posts/show.html.erb:
```diff
@@ -46,25 +46,41 @@
+              <%= f.text_area :comment, placeholder: 'コメントを入力' %>
+            </div>
+            <div class="comment_area__form_box__form__hidden_field">
+              <%= f.hidden_field :user_id, value: current_user.id %>
+              <%= f.hidden_field :post_id, value: @post.id %>
```
これらは削除して、コントローラーで処理をしましょう

この方法もとても面白いと思いますが、user_idとposts_idをhtml上に表示させてしまうので、セキュリティ面で不安が残ると思います。
また、html上でuser_idとpost_idの数字を変更すれば、「別の投稿に、別のuserで投稿できてしまう」ので、避けるべき手法だと思います。

tattu501 replied:
user_idとpost_idのhidden_fieldを削除しました。

msgyu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,16 @@
+class CommentsController < ApplicationController
+  def create
+    @comment = current_user.comments.new(comment_params)
```
`@comment = @post.comment.new(comment_params)`
で定義しましょう

tattu501 replied:
@comment = @post.comment.new(comment_params)に修正しました。

tattu501 replied:
@comment = @post.comment.new(comment_params)だとエラーが発生したので、修正しました。
修正前）
@comment = @post.comment.new(comment_params)

修正後）
@post = Post.find(params[:post_id])
@comment = @post.comments.new(comment_params)

msgyu commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,16 @@
+
+  private
+    def comment_params
+      params.require(:comment).permit(:comment, :user_id, :post_id)
```
`params.require(:comment).permit(:comment).merge(user_id: current_user.id)`
に変更しましょう

tattu501 replied:
params.require(:comment).permit(:comment).merge(user_id: current_user.id)に変更しました。

msgyu commented on app/controllers/posts_controller.rb:
```diff
@@ -8,6 +8,8 @@ def index
 
   def show
     @post = Post.find(params[:id])
+    @comments = @post.comments.order("created_at DESC")
```
includesをつけましょう
@post.comments.includes(:user)

tattu501 replied:
includes(:user)を付与しました。

tattu501 committed:
app/assets/stylesheets/comments.scssを削除

不要なファイルなので削除

tattu501 committed:
comments_controller.rbにbefore_action :authenticate_user!を追加

ログインユーザのみコメントを投稿できるよう、before_action :authenticate_user!を追加

tattu501 committed:
posts/show.html.erbを修正

user_idとposts_idのhidden_fieldを削除

tattu501 committed:
comments_controller.rbを修正

以下の通り修正しました。
@comment = current_user.comments.new(comment_params)
→　@comment = @post.comment.new(comment_params)

tattu501 committed:
comments_controller.rbを修正

以下の通り修正しました。
params.require(:comment).permit(:comment, :user_id, :post_id)
→　params.require(:comment).permit(:comment).merge(user_id: current_user.id)

tattu501 committed:
posts_controller.rbを修正

以下の通り修正しました。
@comments = @post.comments.order("created_at DESC")
→　@comments = @post.comments.includes(:user).order("created_at DESC")

tattu501 committed:
comments_controller.rbのcreateアクションを修正

@comment = @post.comment.new(comment_params)にしたところエラーが発生しましたので、再度修正しました。

msgyu commented:
LGTM

masato-sa93 commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,19 @@
+    @post = Post.find(params[:post_id])
+    @comment = @post.comments.new(comment_params)
+    if @comment.save
+      redirect_back(fallback_location: root_path)
```
ここで`redirect_back`を使っている理由はなんでしょうか？

`redirect_back`はRails5.1以降では非推奨のはずです。
https://qiita.com/100010/items/3ae5dea25e85b0cbf1ee

tattu501 replied:
1つ前のページに戻るならredirect_backだと思ったからです。
上記の通り非推奨なので、redirect_toに変更しました。

masato-sa93 commented on app/controllers/posts_controller.rb:
```diff
@@ -8,6 +8,8 @@ def index
   def show
     @post = Post.find(params[:id])
+    @comments = @post.comments.includes(:user).order("created_at DESC")
+    @comment = @post.comments.build(user_id: current_user.id, post_id: @post.id) if current_user
```
`if current_user`の記述で、未ログイン時に投稿詳細ページに遷移したときにエラー出ないでしょうか？
出なければそのままでOKです。

tattu501 replied:
未ログイン時にもエラーが出ませんでした。なのでそのままにしました。

masato-sa93 reviewed (changes requested):
レビュー完了しましたー！
修正お願いします！

tattu501 committed:
comments_controller.rbを修正

redirect_back(非推奨)をredirect_toに修正しました。

masato-sa93 reviewed (approved):
LGTMですー！
早速の修正さすがです！
